File: cuda/data_association.py
import math
import numpy as np
import scipy
import numba as nb
from numba import cuda

def assign(dist, threshold):
    M, N = dist.shape

    tuples = [(i, j, dist[i, j]) for i in range(M) for j in range(N)]
    # maximize
    tuples.sort(key=lambda t: -t[2])
    tuples = [(t[0], t[1]) for t in tuples]

    assigned_a = set()
    assigned_b = set()
    assigned_total = 0
    assignment_lm = {}

    cost = 0
    for i in range(M*N):
        a, b = tuples[i]

        if a in assigned_a or b in assigned_b:
            continue
        elif dist[a, b] > threshold:
            assignment_lm[a] = b

            assigned_total += 1
            assigned_a.add(a)
            assigned_b.add(b)
            cost += dist[a, b]

        if assigned_total == M:
            break

    return assignment_lm, cost


# @nb.njit()
# def get_dist_matrix(landmarks, measurements, landmarks_cov, measurement_cov):
#     M = len(landmarks)
#     N = len(measurements)
#     dist = np.zeros((M, N), dtype=np.float32)

#     for i in range(M):
#         for j in range(N):
#             cov = landmarks_cov[i] + measurement_cov

#             dist[i, j] = pdf_2(landmarks[i], mean=measurements[j], cov=cov)

#     return dist


def find_unassigned_measurement_idx(assignement, n_measurements):
    all_idx = set(np.arange(n_measurements))
    assigned_idx = set(assignement.values())
    unassigned_idx = all_idx.difference(assigned_idx)

    return unassigned_idx

def associate_landmarks_measurements(particle, measurements, measurement_cov, threshold):
    landmarks = particle.landmark_means
    landmarks_cov = particle.landmark_covariances

    # M = len(landmarks)
    # N = len(measurements)

    M = landmarks.shape[0]
    N = measurements.shape[0]

    pos = np.array([particle.x, particle.y])
    measurement_predicted = landmarks - pos

    dist = get_dist_matrix(measurement_predicted, measurements,
                           landmarks_cov, measurement_cov)
    assignment, _ = assign(dist)

    unassigned_measurement_idx = find_unassigned_measurement_idx(assignment, N)

    # particle.add_landmarks(new_landmarks, measurement_cov)
    return assignment, unassigned_measurement_idx



if __name__ == "__main__":
    M = 2
    N = 2
    np.random.seed(0)
    dist = np.random.uniform(0, 1, size=(M, N))
    # np.save("in.npy", dist)
    # dist = np.array([
    #     [1, 2, 10],
    #     [1, 10, 10],
    #     [10, 10, 1]
    # ])

    assignment_lm, assignment_ml, cost = assign(dist)
    print(assignment_lm, assignment_ml)
    print("Cost:", cost)

File: cuda/data_association.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <stdbool.h>

#include "sort.h"

typedef struct 
{
    float *matrix;
    int n_landmarks;
    int n_measurements;
} dist_matrix;

typedef struct 
{
    int *assignment;
    bool *assigned_landmarks;
    bool *assigned_measurements;
} assignment;

float pdf(float *x, float *mean, float* cov);


void assign(dist_matrix *matrix, assignment *assignment, float threshold) {
    int n_landmarks = matrix->n_landmarks;
    int n_measurements = matrix->n_measurements;

    int *landmark_idx = malloc(n_landmarks * n_measurements * sizeof(int));
    int *measurement_idx = malloc(n_landmarks * n_measurements * sizeof(int));

    for(int i = 0; i < n_landmarks; i++) {
        for(int j = 0; j < n_measurements; j++) {
            landmark_idx[i * n_measurements + j] = i;
            measurement_idx[i * n_measurements + j] = j;
        }
    }

    // float *dist_matrix = malloc(n_landmarks * n_measurements * sizeof(float));
    float *matrix_copy = malloc(n_landmarks * n_measurements * sizeof(float));
    for (int i = 0; i < n_landmarks * n_measurements; i++) {
        matrix_copy[i] = matrix->matrix[i]; 
    }

    quicksort(matrix_copy, landmark_idx, measurement_idx, 0, (n_landmarks * n_measurements) - 1);

    // for (int i = 0; i < n_landmarks * n_measurements; i++) {
    //     printf("%f, ", matrix_copy[i]); 
    // }
    // printf("\n");

    free(matrix_copy);

    int assigned_total = 0;
    float cost = 0;

    for(int i = 0; i < n_landmarks * n_measurements; i++) {
        int a = landmark_idx[i];
        int b = measurement_idx[i];

        if(assignment->assigned_landmarks[a] || assignment->assigned_measurements[b]){
            continue;
        }
        
        if(matrix->matrix[a * n_measurements + b] > threshold){
            assignment->assignment[a] = b;
            assignment->assigned_landmarks[a] = true;
            assignment->assigned_measurements[b] = true;
            assigned_total += 1;
            cost += matrix->matrix[a * n_measurements + b];
        }

        if(assigned_total == n_landmarks) {
            break;
        }
    }

    printf("Cost: %f\n", cost);

    free(landmark_idx);
    free(measurement_idx);
}


void compute_dist_matrix(float **landmarks, float **measurements, dist_matrix *matrix, float *landmarks_cov, float *measurement_cov)
{
    for(int i = 0; i < matrix->n_landmarks; i++) {
        for(int j = 0; j < matrix->n_measurements; j++) {
            float cov[4] = { landmarks_cov[i] + measurement_cov[0], landmarks_cov[i+1], landmarks_cov[i+2], landmarks_cov[i+3] + measurement_cov[1] };
            matrix->matrix[i * matrix->n_measurements + j] = pdf(landmarks[i], measurements[j], cov);
        }
    }
}


float pdf(float *x, float *mean, float* cov)
{
    float a = cov[0];
    float b = cov[1];

    float logdet = log(a*a - b*b);

    float root = sqrt(2)/2.0;
    float e = root * (1/sqrt(a-b));
    float f = root * (1/sqrt(a+b));

    float m = x[0] - mean[0];
    float n = x[1] - mean[1];

    float maha = 2*(m*m*e*e + n*n*f*f);
    float log2pi = log(2 * M_PI);
    return exp(-0.5 * (2*log2pi + maha + logdet));
}


int main()
{
    srand(0);

    int n_landmarks = 1000;
    int n_measurements = 100;

    // float m[] = { 1, 2, 10, 3, 6, 22, 4, 6, 5, 6, 3, 2, 1, 99, 2, 2, 7, 8, 1, 6, 33, 44, 17, 11, 31 };

    float *m = malloc(n_landmarks * n_measurements * sizeof(float));
    for(int i = 0; i < n_landmarks * n_measurements; i++) {
        m[i] = (float)rand()/(float)(RAND_MAX);
    }


    dist_matrix *matrix = malloc(sizeof(dist_matrix));
    matrix->matrix = m;
    matrix->n_landmarks = n_landmarks;
    matrix->n_measurements = n_measurements;

    bool *assigned_landmarks = malloc(n_landmarks * sizeof(bool));
    bool *assigned_measurements = malloc(n_measurements * sizeof(bool));

    for(int i = 0; i < n_landmarks; i++) {
        assigned_landmarks[i] = false;
    }

    for(int i = 0; i < n_measurements; i++) {
        assigned_measurements[i] = false;
    }

    int *assignment_lm = malloc(n_landmarks * sizeof(int));
    for(int i = 0; i < n_landmarks; i++) {
        assignment_lm[i] = -1;
    }

    assignment *assignment = malloc(sizeof(assignment));
    assignment->assignment = assignment_lm;
    assignment->assigned_landmarks = assigned_landmarks;
    assignment->assigned_measurements = assigned_measurements;

    assign(matrix, assignment, 0.0);

    // for(int i = 0; i < n_landmarks; i++) {
    //     printf("%d -> %d\n", i, assignment->assignment[i]);
    // }

    // for(int i = 0; i < n_landmarks; i++) {
    //     printf("%d,", assigned_landmarks[i]);
    // }
    // printf("\n");

    // for(int i = 0; i < n_measurements; i++) {
    //     printf("%d,", assigned_measurements[i]);
    // }
    // printf("\n");


    // free(matrix->matrix);
    free(matrix);
    free(assigned_landmarks);
    free(assigned_measurements);
    free(assignment_lm);
    free(assignment);




    // int n = sizeof(arr) / sizeof(arr[0]);
    // quickSort(arr, lm, me, 0, n - 1);
    // printf("Sorted array: \n");
    // printArray(arr, lm, me, n);
    return 0;
}